### PR TITLE
Added Dark Theme

### PR DESCRIPTION
**Pull request by E5TR**
## Overview
Hello, I added an option to switch between a light and dark theme, and saves preferences in a settings file.
I had to copy dark theme for the css from https://github.com/mrdrivingduck/github-markdown-css and format it so it could work with the class names from the markdown html. I also improved readibility of main.js :)

(This is my first pull request to a repo I'm not directly involved with, so I'm sorry if this isn't good)

## Dark theme:
image
image

## Light theme:
image
image

### CODE REVIEW

**Comment by BlankFaces** [comment]
Once I have enough spare time, I'll look into having multiple themes, and having the current dark theme more consistant, as its using dracula colour scheme for highlighting but not for the markdown itself. 


There are some limitations with what I have implimented, which is that there can only be one light theme and one dark theme, this is mu first time working with electron so I'll have to research it more to find a better way of implimenting it.


I feel like the current state is still good, as at minimum there is a dark theme and light theme now but can still be improved.


I hope this helps in any way :)

**Comment by BlankFaces** [comment]
I fixed the inconsistancies within the dark theme now :)

File: app/index.html [hunk 1]
```diff
@@ -3,9 +3,14 @@
...
         <meta http-equiv="Content-Security-Policy" content="script-src 'self'"/>
-        <link rel="stylesheet" href="../node_modules/highlight.js/styles/vs.css" />
-        <link rel="stylesheet" href="../node_modules/github-markdown-css/github-markdown.css" />
         <link rel="stylesheet" href="index.css" />
+        <style>
+            @import "../node_modules/github-markdown-dracula-css/github-markdown-dracula.css" screen and (prefers-color-scheme: dark);
+            @import "../node_modules/highlight.js/styles/dracula.css" screen and (prefers-color-scheme: dark);
+
+            @import "../node_modules/github-markdown-css/github-markdown.css" screen and (prefers-color-scheme: light);
+            @import "../node_modules/highlight.js/styles/vs.css" screen and (prefers-color-scheme: light);
```
Comment (c3er) [inline app/index.html:12]: Did you consider the print feature? At first glance it does not look like that but I didn't try this yet.

File: app/main.js [hunk 2]
```diff
@@ -5,12 +5,14 @@ const path = require("path")
 const url = require("url")
 
 const childProcess = require("child_process")
-const electron = require("electron")
+const { app, ipcMain, nativeTheme, dialog, Menu, BrowserWindow, session } = require("electron")
```
Comment (c3er) [inline app/main.js:8]: Please revert this change. I "required" all the modules this way intentionally to keep all the content of the modules in their own namespace respectively.

Reasons:

1) To not have to much identifiers in the global namespace
2) To know at first glance where functions, objects etc. belong to

File: app/main.js [hunk 3]
```diff
@@ -5,12 +5,14 @@ const path = require("path")
...
 
 const common = require("./lib/common")
 const encodingStorage = require("./lib/encodingStorage")
 const ipc = require("./lib/ipc")
 
+const storage = require('electron-json-storage');
+
```
Comment (c3er) [inline app/main.js:15]: The requires have a particular order:

```js
// Modules that are available for all Node applications
const fs = require("fs")
const path = require("path")
const url = require("url")

// Libraries that are included via package.json
const childProcess = require("child_process")
const electron = require("electron")

// Own libraries
const common = require("./lib/common")
const encodingStorage = require("./lib/encodingStorage")
const ipc = require("./lib/ipc")

// Each section is ordered alphabetically respectively
```

File: package.json [hunk 4]
```diff
@@ -65,11 +65,15 @@
     ]
   },
   "dependencies": {
+    "electron-json-storage": "^4.4.0",
     "github-markdown-css": "4.0.0",
+    "github-markdown-dracula-css": "^4.0.2",
```
Comment (c3er) [inline package.json:70]: Look out for the syntax of the package versions. I require always *the exact* version to be (almost) sure that any commit/version can be built in the future - maybe even in years to come.

**Review by c3er — CHANGES_REQUESTED** [review]

**Comment by c3er** [comment]
The first pull request I got! I guess, it's time to add a CONTRIBUTING file.

> I added an option to switch between a light and dark theme, and saves preferences in a settings file

Awesome! I was thinking about that feature but didn't have any pressure for it yet (I have my IDEs set to light theme).

I looked at the changes only superficially yet. I will look deeper at this at the weekend. But I added already some comments for you to work on. I don't think that the readability of the `main.js` is improved - see the respective review comment.

If your changes are okay at the weekend so far, I will publish a new release "1.5.0". Currently I have only at weekends time to occupy myself with this project.

> (This is my first pull request to a repo I'm not directly involved with, so I'm sorry if this isn't good)

I see no reason to be sorry🙂 But I suggest the next time you want to add a feature to a project, you open a proposal as issue. For the case that the author just doesn't want that feature and you don't waste any time.

**Comment by c3er** [comment]
> Once I have enough spare time, I'll look into having multiple themes, and having the current dark theme more consistant, as its using dracula colour scheme for highlighting but not for the markdown itself.
> 
> There are some limitations with what I have implimented, which is that there can only be one light theme and one dark theme, this is my first time working with electron so I'll have to research it more to find a better way of implimenting it.
> 
> I feel like the current state is still good, as at minimum there is a dark theme and light theme now but can still be improved.

I think, a dark theme is already a major improvement and would increase the user base dramatically.

If you want to add support for additional themes, you should clearly open another pull request. Maybe you should open a proposal-issue first where this feature and its user interface can be discussed.

File: app/main.js (hunk 2 above)
Comment (BlankFaces) [reply app/main.js]: I have done so now, ( it fixed a bug originally too, to do with nativeTheme originally but when I reverted it somehow fixed the original issue )

File: app/main.js (hunk 3 above)
Comment (BlankFaces) [reply app/main.js]: Fixed in new push :)

File: package.json (hunk 4 above)
Comment (BlankFaces) [reply package.json]: Thank you! I have changed this also

**Comment by BlankFaces** [comment]
> 
> 
> The first pull request I got! I guess, it's time to add a CONTRIBUTING file.
> 
> > I added an option to switch between a light and dark theme, and saves preferences in a settings file
> 
> Awesome! I was thinking about that feature but didn't have any pressure for it yet (I have my IDEs set to light theme).
> 
> I looked at the changes only superficially yet. I will look deeper at this at the weekend. But I added already some comments for you to work on. I don't think that the readability of the `main.js` is improved - see the respective review comment.
> 
> If your changes are okay at the weekend so far, I will publish a new release "1.5.0". Currently I have only at weekends time to occupy myself with this project.
> 
> > (This is my first pull request to a repo I'm not directly involved with, so I'm sorry if this isn't good)
> 
> I see no reason to be sorry🙂 But I suggest the next time you want to add a feature to a project, you open a proposal as issue. For the case that the author just doesn't want that feature and you don't waste any time.


Oh okay, I will do that next time, thank you! I have completed all but one request, as I'll have to look into [this](https://github.com/c3er/mdview/pull/12#discussion_r582018605)


> 
> 
> > Once I have enough spare time, I'll look into having multiple themes, and having the current dark theme more consistant, as its using dracula colour scheme for highlighting but not for the markdown itself.
> > There are some limitations with what I have implimented, which is that there can only be one light theme and one dark theme, this is my first time working with electron so I'll have to research it more to find a better way of implimenting it.
> > I feel like the current state is still good, as at minimum there is a dark theme and light theme now but can still be improved.
> 
> I think, a dark theme is already a major improvement and would increase the user base dramatically.
> 
> If you want to add support for additional themes, you should clearly open another pull request. Maybe you should open a proposal-issue first where this feature and its user interface can be discussed.


I possibly may create a proposal-issue, but most likely not (well for a few months at least) just because of the influx of work soon

File: app/index.html (hunk 1 above)
Comment (BlankFaces) [reply app/index.html:12]: Yeah it doesn't work :/

Comment (BlankFaces) [reply app/index.html]: image

Is this okay?

File: app/main.js (hunk 2 above)
Comment (BlankFaces) [reply app/main.js]: I tried running it on my laptop, and didn't throw any errors, ( never installed it there originally )

File: app/index.html (hunk 1 above)
Comment (BlankFaces) [reply app/index.html]: I'm assuming it should only print out in light theme to not waste ink, unless it's to be printed into a pdf, I will have to look into another method

Comment (BlankFaces) [reply app/index.html]: This has made some theming errors, need to fix
image

Comment (c3er) [reply app/index.html]: The PDF looks good in the screenshot.

> I'm assuming it should only print out in light theme to not waste ink, unless it's to be printed into a pdf, if so I will have to look into another method

I don't think that a dark theme makes any sense on paper. The tool cannot and probably should not distinguish between printing on PDF and actual printing. I think "dark PDFs" are quite uncommon.

Comment (BlankFaces) [reply app/index.html]: Okay cool! I'm going to have to add some js to index.html as I will be unable to load in the dracula highlight.js theme properly while importing the vs highlight.js theme as a default

File: package.json (hunk 4 above)
Comment (BlankFaces) [reply package.json]: I removed an accedently added package, it should all be okay now

File: app/main.js [hunk 5]
```diff
@@ -179,6 +181,24 @@ function createWindow() {
...
+        if (hasKey) {
+            storage.get('settings', function(error, object) {
+                if (error) {
+                    console.log('d2')
+                    electron.nativeTheme.themeSource = 'dark'
+                } else {
+                    electron.nativeTheme.themeSource = object.theme;
+                }
+            });
+        }
```
Comment (c3er) [inline app/main.js:199]: Some things I just noticed:

- Anonymous functions should be defined with arrow syntax preferably. So

  ```js
  storage.get('settings', function(error, object) {
      // ...
  }
  ```

  becomes

  ```js
  storage.get('settings', (error, object) => {
      // ...
  }
  ```

- Please write something meaningful in the log messages, e.g. the error object itself.
- Omitting semicolons is preferred in this project. The semicolon in line 188 is even not quite right.

**Review by c3er — CHANGES_REQUESTED** [review]

File: app/main.js (hunk 5 above)
Comment (BlankFaces) [reply app/main.js:197]: Completed!

**Comment by c3er** [comment]
Hi,

is there a chance that you solve the remaining problem in the next weeks or so? The highlighting of certain source code was not quite right yet. And as I just tried the tool with your changes I noticed that you put the "Switch Theme" item in the File menu. I think the View menu is a better fit.

If you say, you won't have time anytime soon, I'd finish this feature by myself in the next weeks (I got some time again). I would prioritize this feature over bug #11 and the architectural changes that I want to do.

BTW: you already did a great work and I'm looking forward to publish a new release with this feature!

**Comment by c3er** [comment]
Your commits are merged, the remaining problems are fixed and a new [release 1.5.0](https://github.com/c3er/mdview/releases/tag/v1.5.0) is published.

Thanks!

**Comment by E5TR** [comment]
> Your commits are merged, the remaining problems are fixed and a new [release 1.5.0](https://github.com/c3er/mdview/releases/tag/v1.5.0) is published.
> 
> Thanks!

Oh nice!! Thank you! I thought I replied to your last comment but I didn't, sorry!
